test/unit/internals/top_bits.cpp

```diff
@@ -0,0 +1,389 @@
+};
+
+auto top_bits_count_true = []<typename T>(auto& runtime, bool verbose, auto const&, T)
+{
+  test_over_top_bits<T>([&](auto x) {
+    std::ptrdiff_t expected = 0;
+    for (int i = 0; i != x.static_size; ++i) expected += x.get(i);
+
+    top_bits mmask{x};
+    TTS_EQUAL(expected, eve::detail::count_true(mmask));
+  });
+};
+
+EVE_TEST_BED("Check first_true(top_bits)"
+            , eve::test::simd::all_types
+            , eve::test::generate(eve::test::no_data)
+            , top_bits_first_true
+            );
+EVE_TEST_BED("Check count_true(top_bits)"
+            , eve::test::simd::all_types
+            , eve::test::generate(eve::test::no_data)
+            , top_bits_count_true
+            );
+
+//==================================================================================================
+// Check the behavior of top_bits over ignore masks
+//==================================================================================================
+auto top_bits_ignore = []<typename T>(auto& runtime, bool verbose, auto const&, T)
+{
+  using logical = eve::logical<T>;
+
+  // ignore all
+  {
+    top_bits<logical> mmask(eve::ignore_all);
+    TTS_EQUAL(logical(false), eve::detail::to_logical(mmask));
+  }
+
+  // ignore none
+  {
+    top_bits<logical> mmask(eve::ignore_none);
+    TTS_EQUAL(logical(true), eve::detail::to_logical(mmask));
+  }
+
+  // ignore_extrema / keep_between
+  {
+    for (int i = 0; i < logical::static_size + 1; ++i)
+    {
+      for (int j = logical::static_size - i; j ; --j)
+      {
+        logical expected([&](int k, int) { return (k >= i) && (logical::static_size - k) > j; });
+        top_bits<logical> actual(eve::ignore_extrema(i, j));
+        TTS_EQUAL(expected, eve::detail::to_logical(actual));
+
+        actual = top_bits<logical>(eve::keep_between(i, logical::static_size - j));
+        TTS_EQUAL(expected, eve::detail::to_logical(actual));
+      }
+    }
+  }
+
+  // ignore first / keep_last
+  {
+    for (int i = 0; i < logical::static_size + 1; ++i)
+    {
+      logical expected([&](int j, int) { return j >= i; });
+      top_bits<logical> actual(eve::ignore_first{i});
+      TTS_EQUAL(expected, eve::detail::to_logical(actual));
+
+      actual = top_bits<logical>(eve::keep_last(logical::static_size - i));
+      TTS_EQUAL(expected, eve::detail::to_logical(actual));
+    }
+  }
+
+  // ignore last / keep_first
+  {
+    for (int i = 0; i < logical::static_size + 1; ++i)
+    {
+      logical expected([&](int j, int) { return (logical::static_size - j) > i; });
+      top_bits<logical> actual(eve::ignore_last{i});
+      TTS_EQUAL(expected, eve::detail::to_logical(actual));
+
+      actual = top_bits<logical>(eve::keep_first(logical::static_size - i));
+      TTS_EQUAL(expected, eve::detail::to_logical(actual));
+    }
+  }
+};
+
+EVE_TEST_BED("Check top_bits from ignore_*"
+            , eve::test::simd::all_types
+            , eve::test::generate(eve::test::no_data)
```

Comment:
What I would suggest maybe is to, when you have no generation, just use another macro?
This is a touch awkward I think.

Reply:
Solved